### PR TITLE
Use pytest not nose

### LINKED ISSUE #1984: Investigate the use pytest instead of nose
Pytest differs from nose in that you can just `assert` anything, and when something fails, pytest will apply some voodoo to come up with an error message that makes sense (e.g. show diff of lists if you assert that two lists are equal). This makes writing tests much easier, because you don't need all the `assert_x` functions.

For asserting exceptions, there is a `raises()` function, that can also be used as a context manager. And that's in most cases the only function you ever need in a test.

Some examples:

    assert some_number == 3
    assert some_list == [2, 3, 4]
    raises(ValueError, foo, 3, 4, bar=8)
    with raises(ValueError):
        foo(3, 4, bar)

We can set this up for Travis in a PR and see how we like it. pytest is compatible with unittest and nose, so transition can be smooth. No need to rewrite our existing tests.

### PR DESCRIPTION
**Pull request by birdsarah**
issues: closes #1984

* Test collection number is not the same
* Getting a few failing tests locally.

### CODE REVIEW

File: scripts/travis_install [hunk 1]
```diff
@@ -26,6 +26,7 @@ conda install --yes $DEPS_TRAVIS
 conda config --add channels bokeh
 conda config --add channels javascript   # TODO: move packages to bokeh?
 conda config --add channels mutirri   # TODO: remove when packages are moved to bokeh
+conda config --add channels jrderuiter   # TODO: remove when pytest-cov available on bokeh channel
```
Comment (bryevdv) [inline scripts/travis_install]: should do before merging the PR

Comment (birdsarah) [inline scripts/travis_install]: Definitely, also on py33 got this error in travis:

````sh
Hint: the following packages conflict with each other:
  - bokeh 0.9.2.73283141 np19py33_1
  - pytest-cov
````

**Comment by birdsarah** [comment]
fyi, the latest commit ensures that we see the full js test output, by running the js test seperately first e.g. https://travis-ci.org/bokeh/bokeh/jobs/73285009#L1100

File: scripts/travis_install (hunk 1 above)
Comment (damianavila) [inline scripts/travis_install]: https://anaconda.org/hargup/pytest-cov/files, seems to have a noarch package... we can probably test that or build one for py33

File: .travis.yml [hunk 2]
```diff
@@ -41,7 +41,8 @@ install:
 script:
     # Deactivate tests on building to avoid huge packages contaning the pics and outout html
     # Eventually, we need to get rid of the generated stuff before packaging
-    - if [[ -z "$TRAVIS_TAG" ]]; then nosetests --with-coverage --cover-package=bokeh; fi
+    - if [[ -z "$TRAVIS_TAG" ]]; then py.test -s -m js; fi  # Run just the JS first so we see the output
```
Comment (damianavila) [inline .travis.yml]: great!

**Comment by birdsarah** [comment]
I re-based as this was getting pretty messy.

I have confirmed that if the JS tests fail but the python tests pass we get a fail on Travis - https://travis-ci.org/bokeh/bokeh/jobs/73970043

@damianavila has kindly offered to upload pytest-cov to bokeh channel, and then this should be good to go.

**Comment by bryevdv** [comment]
Py33 test failing with:
```
Hint: the following packages conflict with each other:
  - bokeh 0.9.2.73974621 nppy33_1
  - pytest-cov
```

**Comment by bryevdv** [comment]
@birdsarah I really like being able to do `bokeh.test()` is that not something that can be kept?

File: conda.recipe/meta.yaml [hunk 3]
```diff
@@ -51,7 +51,8 @@ test:
     - pdiff
     - boto                 [py2k]
     - jinja2
-    - nose
+    - pytest
+    - pytest-cov   # Available in jrderuiter channel
```
Comment (damianavila) [inline conda.recipe/meta.yaml]: It is in the bokeh channel now...

File: scripts/travis_install (hunk 1 above)
Comment (damianavila) [inline scripts/travis_install]: Remove that line, the packages are in the bokeh channel now: https://anaconda.org/bokeh/pytest-cov/files

**Comment by damianavila** [comment]
>I re-based as this was getting pretty messy.

Please, @bokeh/dev, do not rebase after publishing your branch.
For new people, we heavily discuss this in the past and we agreed on a merge-based strategy without rebases after your branch in published.

This remembers me that we need to explicitly say this on a written form in some guidelines to contribute, I guess...

**Comment by damianavila** [comment]
@birdsarah, with the pytest-cov already built and uploaded in the bokeh channel, I restarted the py33 build, it should turn to green soon...

**Comment by fpliger** [comment]
> @birdsarah I really like being able to do bokeh.test() is that not something that can be kept?

@bryevdv @birdsarah Is anything preventing us to use `pytest.main()` instead of `nose.main()'?

**Comment by almarklein** [comment]
pytest should be compatible with nose; so you should always be able to run `pytest.main()` tho there could be some subtleties that I am overlooking.

File: scripts/travis_install (hunk 1 above)
Comment (birdsarah) [inline scripts/travis_install]: Thanks @damianavila

**Comment by birdsarah** [comment]
> @birdsarah I really like being able to do bokeh.test() is that not something that can be kept?
> pytest should be compatible with nose; so you should always be able to run pytest.main() tho there could be some subtleties that I am overlooking.

Yes `bokeh.test()` can be kept (with a few subtleties, but yes). However, I don't think it should be.

The test function will look like this:
````pyth
def test():
    import pytest
    pytest.main()
````
It does nothing except the same as calling py.test from the command line. I don't think this is a good idea because:
1) You have two ways of doing the same thing which leads to confusion - "which one is the right one?", "does bokeh.test() do special magic?"
2) When it comes to configuring test stuff in the future, do you do it in this method or in the normal pytest setup places - pytest.ini or setup.cfg
3) It's just more code to maintain
4) Using py.test from the command line fits with the pytest documentation which makes it clearer for people new to pytest and familiar for people used to pytest.

However, I know that old habits die hard, so I'd be fine with putting it back in with a deprecation warning.

ping @bryevdv @almarklein @fpliger

**Comment by birdsarah** [comment]
Sorry about the rebase @damianavila :person_frowning:

**Comment by almarklein** [comment]
Wouldn't it look more something like this:

    def test():
        import pytest
        pytest.main(BOKEH_ROOT_DIR)

One reason to start the tests from within Python is that one can use post-mortem debugging in your IDE to jump right to the point of failure and inspect local variables etc.. I find this very convenient (I contributed a patch to py.test some time ago to allow this).

Having a function `test()` that makes the path be correct (and maybe set up coverage?) is convenient in that case. We can document that there is no difference from running py.test from the command line.

**Comment by birdsarah** [comment]
@damianavila - there appears to be something wrong with the pytest-cov on bokeh channel

For all versions (py27, py33, py34) on the examples tests, travis is showing : 

````sh
>>> Testing /home/travis/build/bokeh/bokeh/examples/plotting/file/logplot.py ...
Error processing line 1 of /home/travis/miniconda/lib/python3.3/site-packages/pytest-cov.pth:
  Traceback (most recent call last):
    File "<string>", line 3, in <module>
  ImportError: No module named 'pytest_cov'
  During handling of the above exception, another exception occurred:
  Traceback (most recent call last):
    File "/home/travis/miniconda/lib/python3.3/site.py", line 164, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
    File "<string>", line 10, in <module>
  NameError: name 'exc' is not defined
Remainder of file ignored
[OK] /home/travis/build/bokeh/bokeh/examples/plotting/file/logplot.py
````
It doesn't do this on my local machine when I `pip install pytest-cov`. Let me know if I can supply more info.

There could be a workaround - to run the examples tests separately without invoking coverage (which may be more correct actually).

**Comment by almarklein** [comment]
Can we not just `pip install pytest-cov` on Travis?

**Comment by birdsarah** [comment]
@almarklein 

> Wouldn't it look more something like this...

Firstly, I wasn't trying to be precise. I will be more clear next time.

It may look like that, additionally we may want to convolute it further by allowing additional pytest arguments to be passed to it (previously runtests let you pass in the verbosity argument).

> One reason to start the tests from within Python is that one can use post-mortem debugging in your IDE to jump right to the point of failure and inspect local variables etc.. I find this very convenient

I don't use an IDE so I'm not sure about your particular use case. Personally, when I want to jump to the point of a failure I just run `py.test --pdb` and when the test fails I'm immediately in the debugger. 

I don't feel so strongly if there's a real need (as opposed to just habit). If that's the case, we should agree what the set of arguments we want to be able to pass through bokeh.test to py.test are.

**Comment by almarklein** [comment]
> I don't feel so strongly if there's a real need (as opposed to just habit).

I suppose the need is not so high, but it's a matter of taking into account different workflows (i.e. habits) :)

Here's a more realistic take on what test() could look like (based on what I know from other projects):

    def test(verbose=True, cov_report='term')
        os.chdir(ROOT_DIR)
        _enable_faulthandler()  # to get more info for segfaults on py3k
        verbose = '-v' if verbose else ''
        pytest.main('%s -x --color=yes --cov bokeh '
                '--cov-config .coveragerc --cov-report %s tests' %  (verbose, cov_report))

Also note that coverage reports will be incorrect for any modules that are already loaded when this function is called. If this is a problem, the function can remove the modules from `sys.modules` before running the tests.

**Comment by bryevdv** [comment]
I use `bokeh.test()` constantly, and it's common and familiar from other projects. I am in favor of keeping it.

**Comment by damianavila** [comment]
@birdsarah there is some problems with the pytest-cov packages... I explored other recipes and I can see they are pretty similar... I guess there is some issue with the latest pytest-cov, I will try to build the previous version and upload it to the channel (or can really try to install it with pip temporarily) whereas we ping someone from the conda team to take a look on the problem with the latest version... in fact, ping @asmeurer, see comment https://github.com/bokeh/bokeh/pull/2637#issuecomment-127781614 (I create the recipe with conda skeleton, any idea where could be the problem? thanks!)

**Comment by damianavila** [comment]
@birdsarah, I have uploaded pytest-cov 1.8.1 to the bokeh anaconda user (I also have to build some other package to built it correctly). Still not the best, because we don't have the latest version, but I think is enough for now... until we figured out why the 2.0 is no working... btw, looking at TravisCI now, it seems to work...

**Comment by birdsarah** [comment]
@mistakevin @bryevdv @damianavila - I was thinking about splitting up running coverage on the examples tests. They're integration tests, they're going to always plough through a whole bunch of machinery. Are we getting useful coverage numbers by including them, or should we only run coverage with the tests in the main bokeh package.  I did some analysis locally which I can share if you want, but the headline is: the examples tests alone give a coverage of 33%, the package tests alone give a coverage of 65%.

The modules the examples test better are notebook, and widgets (perhaps unsurpringly). This is not line by line analysis, just a quick look.

I can see arguments for not running coverage for the examples tests, running coverage for the examples tests seperately, **but I think that lumping them together hides how are unit tests are doing.**

**Comment by damianavila** [comment]
If we are running coverage over the examples (not sure about that). I think the best option is:

>running coverage for the examples tests separately

**Comment by fpliger** [comment]
> If we are running coverage over the examples (not sure about that). I think the best option is:

>>running coverage for the examples tests separately

+1

**Comment by birdsarah** [comment]
I think this is ready now. Tests run as follows:
* javascript tests (with output from js) - no coverage
* main python unit tests  - coverage
* examples tests - coverage

and......`bokeh.test()` lives again :D

**Comment by almarklein** [comment]
LGTM

**Comment by bryevdv** [comment]
LGTM too, @bokeh/dev I plan to merge this tomorrow unless there is further comment or objection.

**Comment by damianavila** [comment]
LGTM too...

**Pull request merged**